example-crates/wallet_esplora_async/src/main.rs

```diff
@@ -24,62 +33,93 @@ async fn main() -> Result<(), anyhow::Error> {
-    let (update_graph, last_active_indices) = client
-        .full_scan(keychain_spks, STOP_GAP, PARALLEL_REQUESTS)
-        .await?;
-    let missing_heights = update_graph.missing_heights(wallet.local_chain());
-    let chain_update = client.update_local_chain(prev_tip, missing_heights).await?;
-    let update = Update {
-        last_active_indices,
-        graph: update_graph,
-        chain: Some(chain_update),
-    };
+    info!("Wallet balance: {} sats", balance.total());
+
+    let client = esplora_client::Builder::new("http://signet.bitcoindevkit.net").build_async()?;
+
+    let (update, cmd) = match cmd.map(|c| c.as_str()) {
+        Some(cmd) if cmd == "fullscan" => {
+            info!("Start full scan...");
+            // 1. get data required to do a wallet full_scan
+            let request = wallet.full_scan_request();
+            // 2. full scan to discover wallet transactions
+            let (graph_update, last_active_indices) = client
+                .full_scan(request.spks_by_keychain, STOP_GAP, PARALLEL_REQUESTS)
+                .await?;
+            // 3. from wallet transactions determine missing blockchain heights
+            let missing_heights = graph_update.missing_heights(wallet.local_chain());
+            // 4. get blockchain update from original request checkpoint and missing heights
+            let chain_update = client
+                .update_local_chain(request.checkpoint, missing_heights)
+                .await?;
+            // 5. create wallet update
+            Ok((
+                Update {
+                    last_active_indices,
+                    graph: graph_update,
+                    chain: Some(chain_update),
+                },
+                cmd,
+            ))
+        }
+        Some(cmd) if cmd == "sync" => {
+            info!("Start sync...");
+            // 1. get data required to do a wallet sync, if also syncing previously used addresses set unused_spks_only = false
+            let request = wallet.sync_request(true);
+            // 2. sync unused wallet spks (addresses), unconfirmed tx, and utxos
+            let graph_update = client
+                .sync(
+                    request.spks,
+                    request.txids,
+                    request.outpoints,
+                    PARALLEL_REQUESTS,
+                )
+                .await?;
+            // 3. from wallet transactions determine missing blockchain heights
+            let missing_heights = graph_update.missing_heights(wallet.local_chain());
+            // 4. get blockchain update from original request checkpoint and missing heights
+            let chain_update = client
+                .update_local_chain(request.checkpoint, missing_heights)
+                .await?;
+            // 5. create wallet update
+            Ok((
+                Update {
+                    graph: graph_update,
+                    chain: Some(chain_update),
+                    ..Update::default()
```

Comment:
I was thinking about creating some sort of builder structure for ChangeSets. Maybe we can do something similar for Updates?

```rust
pub struct UpdateBuilder<U> { u: U }

impl<U> UpdateBuilder<U> {
    pub fn chain(self, u: impl Into<U>) -> Self { todo!() }
    pub fn build(self) -> U { self.u }
}
```

However changesets can be appended, but updates may not? Maybe `chain` needs to return a result. Would this even be a good idea?

Reply:
I think the default struct construction above is pretty clear and easy to use.

Reply:
Builder pattern would make sense for building a sync/scan request. I think the current approach leaves the door open to potentially evolve in that direction.

Reply:
Why would you need a builder pattern for changesets when you should never create these yourself?
It could be helpful for updates but I don't see what's wrong with the code above?

Oh except I wonder if we could have the concept of the "empty chain" which just has the geneisis block or whatever rather than having to have an `Option` there.

Reply:
📌 Resolving this for now without adding a builder since that can be done later and this PR already has a lot to review.